api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQueryIncludeImpl.java

```diff
@@ -11,23 +11,71 @@
 
 import java.util.HashSet;
 import java.util.List;
+import java.util.Objects;
 import java.util.Set;
 
+import ca.uhn.fhir.model.api.Include;
 import lombok.NoArgsConstructor;
 import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.Location;
+import org.hl7.fhir.r4.model.Reference;
+import org.openmrs.module.fhir2.FhirConstants;
+import org.openmrs.module.fhir2.api.FhirLocationService;
+import org.openmrs.module.fhir2.api.search.param.PropParam;
 import org.openmrs.module.fhir2.api.search.param.SearchParameterMap;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
+import org.springframework.util.CollectionUtils;
 
 @Component
 @NoArgsConstructor
 public class SearchQueryIncludeImpl<U extends IBaseResource> implements SearchQueryInclude<U> {
 	
+	@Autowired
+	private FhirLocationService locationService;
+	
 	@Override
+	@SuppressWarnings("unchecked")
 	public Set<IBaseResource> getIncludedResources(List<U> resourceList, SearchParameterMap theParams) {
 		Set<IBaseResource> includedResourcesSet = new HashSet<>();
 		
-		// TODO -> Add implementation to compute the included resources
+		List<PropParam<?>> includeParamList = theParams.getParameters(FhirConstants.INCLUDE_SEARCH_HANDLER);
+		
+		if (CollectionUtils.isEmpty(includeParamList)) {
+			return includedResourcesSet;
+		}
+		
+		Set<Include> includeSet = (HashSet<Include>) includeParamList.get(0).getParam();
+		includeSet.forEach(includeParam -> {
+			switch (includeParam.getParamName()) {
+				case FhirConstants.INCLUDE_PART_OF_PARAM:
+					includedResourcesSet.addAll(handlePartofInclude(resourceList, includeParam.getParamType()));
+					break;
+			}
+		});
 		
 		return includedResourcesSet;
 	}
+	
+	private Set<IBaseResource> handlePartofInclude(List<U> resourceList, String paramType) {
+		Set<IBaseResource> includedResources = new HashSet<>();
+		Set<String> uniqueParentLocationUUIDs = new HashSet<>();
+		
+		switch (paramType) {
+			case FhirConstants.LOCATION:
+				resourceList.forEach(
+				    resource -> uniqueParentLocationUUIDs.add(getIdFromReference(((Location) resource).getPartOf())));
+				
+				uniqueParentLocationUUIDs.removeIf(Objects::isNull);
+				includedResources.addAll(locationService.get(uniqueParentLocationUUIDs));
+				break;
+		}
+		
+		return includedResources;
+	}
```

Comment:
This seems unnecessarily verbose. Couldn't this whole thing be simplified to something like:

```
private static Set<Location> handlePartofInclude(List<Location> locations) {
	resourceList.stream().map(Location::getPartOf).filter(Objects::nonNull).map(ref -> getIdFromReference(ref)).collect(Collectors.toSet());
}
```

I.e., do we actually need a setup similar to what we do for searching (where the point is to be able to combine a series of operations to produce a single Criteria object for running the query) or can we get away with something much simpler?

Reply:
No, so the thing here is that we could have partof for multiple resources in the future. Its use is not clearly evident here but let us say we support _include=Location:partof as well as _include=Observation:partof, instead of creating a separate function for each of these, we can just club them together based on the resource type.

Reply:
This would look something like - 
```
	private Set<IBaseResource> handlePatientInclude(List<U> resourceList, String paramType) {
		Set<IBaseResource> includedResources = new HashSet<>();
		Set<String> uniquePatientUUIDs = new HashSet<>();
		
		switch (paramType) {
			case FhirConstants.OBSERVATION:
				resourceList.forEach(
				    resource -> uniquePatientUUIDs.add(getIdFromReference(((Observation) resource).getSubject())));
				break;
			case FhirConstants.ALLERGY_INTOLERANCE:
				resourceList.forEach(
				    resource -> uniquePatientUUIDs.add(getIdFromReference(((AllergyIntolerance) resource).getPatient())));
				break;
			case FhirConstants.DIAGNOSTIC_REPORT:
				resourceList.forEach(
				    resource -> uniquePatientUUIDs.add(getIdFromReference(((DiagnosticReport) resource).getSubject())));
				break;
			case FhirConstants.ENCOUNTER:
				resourceList.forEach(
				    resource -> uniquePatientUUIDs.add(getIdFromReference(((Encounter) resource).getSubject())));
				break;
			case FhirConstants.PERSON:
				resourceList.forEach(resource -> {
					List<Reference> patientReferenceList = new ArrayList<>();
					((Person) resource).getLink().forEach(patient -> patientReferenceList.add(patient.getTarget()));
					uniquePatientUUIDs.addAll(getIdsFromReferenceList(patientReferenceList));
				});
				break;
			case FhirConstants.RELATED_PERSON:
				resourceList.forEach(resource -> uniquePatientUUIDs.add(getIdFromReference(((RelatedPerson) resource).getPatient())));
				break;
		}
		
		uniquePatientUUIDs.removeIf(Objects::isNull);
		includedResources.addAll(patientService.get(uniquePatientUUIDs));
		
		return includedResources;
	}
```

Reply:
So, I can get that there are cases where this pattern can be useful, but in the specific case of `Location:partof` I don't think it is. I think, like with your patient example, it makes sense to group the includes into functions *based on the result type rather than the search query*.

For instance, if we had a `handlePartOf()` method that handles both Location and Observation, there's essentially no logic shared between those paths. But it would make sense to have a `handleLocation()` function that takes various parameters to extract Location UUIDs and then runs a single `locationService.get()` call.

Reply:
Incidentally, functions based on the result type are essentially what your `handlePatientInclude()` does.